tutorial/source/effect_handlers.ipynb

```diff
@@ -0,0 +1,629 @@
+    "\n",
+    "Finally, `TraceMessenger` and `ConditionMessenger` are Pyro effect handlers, or `Messenger`s: stateful context manager objects that are placed on a global stack and pass messages (hence the name) up and down the stack at each effectful operation, like a `pyro.sample` call.  A `Messenger` is placed at the bottom of the stack when its `__enter__` method is called, i.e. when it is used in a \"with\" statement.\n",
+    "\n",
+    "For a simplified implementation of this mechanism in only a few lines of code, see [pyro.contrib.minipyro](https://github.com/uber/pyro/blob/dev/pyro/contrib/minipyro.py). For background reading on effect handlers, see the paper [\"Handlers in Action\"](http://homepages.inf.ed.ac.uk/slindley/papers/handlers.pdf) by Ohad Kammar, Sam Lindley, and Nicolas Oury."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Implementing new effect handlers with the `Messenger` API\n",
+    "\n",
+    "Although it's easiest to build new effect handlers by composing the existing ones in `pyro.poutine`, implementing a new effect as a `pyro.poutine.messenger.Messenger` subclass is actually fairly straightforward. As a first example, let's implement a version of our log-joint computation that performs the sum while the model is executing."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "tensor(-3.0203)\n",
+      "tensor(-3.0203)\n"
+     ]
+    }
+   ],
+   "source": [
+    "class LogJointMessenger(poutine.messenger.Messenger):\n",
+    "    \n",
+    "    def __init__(self, data):\n",
+    "        self.data = data\n",
+    "    \n",
+    "    # __call__ allows Messengers to be used as higher-order functions.\n",
+    "    # Messenger already defines __call__, but we re-define it here\n",
+    "    # for exposition and to change the return value:\n",
+    "    def __call__(self, fn):\n",
+    "        def _fn(*args, **kwargs):\n",
+    "            with self:\n",
+    "                fn(*args, **kwargs)\n",
+    "                return self.logp.clone()\n",
+    "        return _fn\n",
+    "    \n",
+    "    def __enter__(self):\n",
+    "        self.logp = torch.tensor(0.)\n",
+    "        return super(LogJointMessenger, self).__enter__()\n",
+    "    \n",
+    "    # __exit__ takes the same arguments in all Python context managers\n",
+    "    def __exit__(self, exc_type, exc_value, traceback):\n",
+    "        self.logp = torch.tensor(0.)\n",
+    "        return super(LogJointMessenger, self).__exit__(exc_type, exc_value, traceback)\n",
+    "        \n",
+    "    def _pyro_sample(self, msg):\n",
+    "        assert msg[\"name\"] in self.data\n",
+    "        msg[\"value\"] = self.data[msg[\"name\"]]\n",
+    "        msg[\"is_observed\"] = True\n",
```

Comment:
what's going on here?

`msg["is_observed"] = True`

Reply:
is the necessary and if not isn't it confusing?

Reply:
It's only necessary as currently implemented - otherwise the value gets overridden with a sample by `default_process_message`.  I'll change that though since it's a confusing edge case.